# EDA

- 탐색적 데이터 분석(Exploratory Data Analysis)
- 데이터 세트를 분석하고 이해하는 과정에서 사용되는 다양한 기술과 절차
    -  데이터의 주요 특성을 파악하고, 데이터 내 숨겨진 패턴, 이상치, 구조 등을 발견하는 것

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# from sklearn.preprocessing import Imputer  # imputer 더 이상 지원안함.
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import VarianceThreshold
from sklearn.feature_selection import SelectFromModel
from sklearn.utils import shuffle
from sklearn.ensemble import RandomForestClassifier

pd.set_option('display.max_columns', 100)

Table of Contents
1. 데이터의 전체적인 특징
2. 메타데이터프레임 만들기
3. 통계 분석
4. 데이터 전처리
5. EDV(Exploratory Data Visualization)
6. Feature Enginering
7. Feature 고르기
8. Feature 스케일링

In [9]:
# 데이터 불러오기
train_df = pd.read_csv(r"N:\Algorithm Study\WEEK14\train.csv")
train_df.head(5)

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0              0              0              1              0              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              0              1              0              0              0   
4              0              1              0              0              0   

   ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  ps_ind_14  \
0              0              0              0              0          0   
1              0              0              0              0          0   
2              0              0              0              0          0   
3              0              0              0              0          0   
4              0              0              0              0          0   

   ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_reg_01  \
0         11              0              1              0        0.7   
1          3              0              0              1        0.8   
2         12              1              0              0        0.0   
3          8              1              0              0        0.9   
4          9              1              0              0        0.7   

   ps_reg_02  ps_reg_03  ps_car_01_cat  ps_car_02_cat  ps_car_03_cat  \
0        0.2   0.718070             10              1             -1   
1        0.4   0.766078             11              1             -1   
2        0.0  -1.000000              7              1             -1   
3        0.2   0.580948              7              1              0   
4        0.6   0.840759             11              1             -1   

   ps_car_04_cat  ps_car_05_cat  ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  \
0              0              1              4              1              0   
1              0             -1             11              1              1   
2              0             -1             14              1              1   
3              0              1             11              1              1   
4              0             -1             14              1              1   

   ps_car_09_cat  ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  \
0              0              1             12          2   0.400000   
1              2              1             19          3   0.316228   
2              2              1             60          1   0.316228   
3              3              1            104          1   0.374166   
4              2              1             82          3   0.316070   

   ps_car_13  ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  \
0   0.883679   0.370810   3.605551         0.6         0.5         0.2   
1   0.618817   0.388716   2.449490         0.3         0.1         0.3   
2   0.641586   0.347275   3.316625         0.5         0.7         0.1   
3   0.542949   0.294958   2.000000         0.6         0.9         0.1   
4   0.565832   0.365103   2.000000         0.4         0.6         0.0   

   ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  \
0           3           1          10           1          10           1   
1           2           1           9           5           8           1   
2           2           2           9           1           8           2   
3           2           4           7           1           8           4   
4           2           2           6   

In [10]:
test_df = pd.read_csv(r'N:\Algorithm Study\WEEK14\test.csv')
test_df.head()

id  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  ps_ind_05_cat  \
0   0          0              1          8              1              0   
1   1          4              2          5              1              0   
2   2          5              1          3              0              0   
3   3          0              1          6              0              0   
4   4          5              1          7              0              0   

   ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  ps_ind_10_bin  \
0              0              1              0              0              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              1              0              0              0              0   
4              0              0              0              1              0   

   ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  ps_ind_14  ps_ind_15  \
0              0              0              0          0         12   
1              0              0              0          0          5   
2              0              0              0          0         10   
3              0              0              0          0          4   
4              0              0              0          0          4   

   ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_reg_01  ps_reg_02  \
0              1              0              0        0.5        0.3   
1              1              0              0        0.9        0.5   
2              0              0              0        0.4        0.0   
3              1              0              0        0.1        0.2   
4              1              0              0        0.9        0.4   

   ps_reg_03  ps_car_01_cat  ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  \
0   0.610328              7              1             -1              0   
1   0.771362              4              1             -1              0   
2   0.916174             11              1             -1              0   
3  -1.000000              7              1             -1              0   
4   0.817771             11              1             -1              0   

   ps_car_05_cat  ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
0             -1              1              1              1              2   
1              0             11              1              1              0   
2             -1             14              1              1              2   
3             -1              1              1              1              2   
4             -1             11              1              1              2   

   ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  ps_car_14  \
0              1             65          1   0.316228   0.669556   0.352136   
1              1            103          1   0.316228   0.606320   0.358329   
2              1             29          3   0.400000   0.896239   0.398497   
3              1             40          2   0.374166   0.652110   0.381445   
4              1            101          3   0.374166   0.812914   0.385097   

   ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  ps_calc_05  \
0   3.464102         0.1         0.8         0.6           1           1   
1   2.828427         0.4         0.5         0.4           3           3   
2   3.316625         0.6         0.6         0.6           2           3   
3   2.449490         0.1         0.5         0.5           2           1   
4   3.316625         0.9         0.6         0.8           3           4   

   ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  ps_calc_10  ps_calc_11  \
0           6           3           6           2           9           1   
1           8           4          10           2           7           2   
2           7           4           6           3          12           4   
3           7        

# 데이터의 특징

- 비슷한 그룹핑에 들어있는 피쳐는 다음과 같이 표시됨
    - ind, reg, car, calc...
- bin 이라고 되어 있는 칼럼은 binary, cat은 고양이가 아니라 카테고리형 데이터임
- 아무런 표시가 없는 데이터 칼럼은 연속적인 데이터임
- 결측치는 NaN이 아니라 -1로 표시됨
- Target 칼럼은 0이면 내년도에 보험들지 않았고, 1이면 보험을 들었음

In [11]:
# pandas에서 사용되는 명령어. DataFrame의 마지막 몇 개의 행(기본적으로 5개)을 반환

train_df.tail()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
595207  1488013       0          3              1         10              0   
595208  1488016       0          5              1          3              0   
595209  1488017       0          1              1         10              0   
595210  1488021       0          5              2          3              1   
595211  1488027       0          0              1          8              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              1              0              0   
595210              0              0              0              1   
595211              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
595207              1              0              0              0   
595208              1              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
595207              0          0         13              1              0   
595208              0          0          6              1              0   
595209              0          0         12              1              0   
595210              0          0         12              1              0   
595211              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
595207              0        0.5        0.3   0.692820             10   
595208              0        0.9        0.7   1.382027              9   
595209              0        0.9        0.2   0.659071              7   
595210              0        0.9        0.4   0.698212             11   
595211              0        0.1        0.2  -1.000000              7   

        ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  ps_car_05_cat  \
595207              1             -1              0              1   
595208              1             -1              0             -1   
595209              1             -1              0             -1   
595210              1             -1              0             -1   
595211              0             -1              0             -1   

        ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
595207              1              1              1              0   
595208             15              0              0              2   
595209              1              1              1              2   
595210             11              1              1              2   
595211              0              1              0              2   

        ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  \
595207              1             31          3   0.374166   0.684631   
595208              1             63          2   0.387298   0.972145   
595209              1             31          3   0.397492   0.596373   
595210              1            101          3   0.374166   0.764434   
595211              1             34          2   0.400000   0.932649   

        ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  \
595207   0.385487   2.645751         0.4         0.5         0.3           3   
595208  -1.000000   3.605551         0.2         0.2         0.0           2   
595209   0.398748   1.732051         0.4         0.0         0.3           3   
595210   0.384968   3.162278         0.0         0.7         0.0           4   
595211   0.378021   3.741657         0.4         0.0         0.5           2   

        ps_calc_05  ps_calc_06  

(중간 요약)

- 첫번째와 마지막 5개의 행만 보고 데이터 대충 판단함

앞으로 체크할 것

- binary 변수
- 정수값으로 구성된 카테고리 변수
- 정수나 실수값으로 구성된 변수
- -1로 된 결측치
- 타겟변수와 id 변수

# 여기서 부터 train_df를 살펴보자

In [12]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 595212 entries, 0 to 595211
Data columns (total 59 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              595212 non-null  int64  
 1   target          595212 non-null  int64  
 2   ps_ind_01       595212 non-null  int64  
 3   ps_ind_02_cat   595212 non-null  int64  
 4   ps_ind_03       595212 non-null  int64  
 5   ps_ind_04_cat   595212 non-null  int64  
 6   ps_ind_05_cat   595212 non-null  int64  
 7   ps_ind_06_bin   595212 non-null  int64  
 8   ps_ind_07_bin   595212 non-null  int64  
 9   ps_ind_08_bin   595212 non-null  int64  
 10  ps_ind_09_bin   595212 non-null  int64  
 11  ps_ind_10_bin   595212 non-null  int64  
 12  ps_ind_11_bin   595212 non-null  int64  
 13  ps_ind_12_bin   595212 non-null  int64  
 14  ps_ind_13_bin   595212 non-null  int64  
 15  ps_ind_14       595212 non-null  int64  
 16  ps_ind_15       595212 non-null  int64  
 17  ps_ind_16_

892816개의 행, 58개의 열로 구성되어 있다. test데이터셋은 target이 존재하지 않으므로 문제가 없어보인다.

Metadata 
효과적으로 데이터프레임을 분석하기 위해서, 우리는 변수들에 관한 메타 정보들을 새롭게 저장할 것이다. 

메타데이터는 우리가 상세한 분석이나 시각화, 모델링 등에 활용될 수 있다. 아래는 데이터의 특성에 따라 4가지 항목으로 분류한 것이다. 

영어로 적는게 더 편할 것 같아 따로 한글로 바꾸진 않았다,

role : input, ID, target

level : nominal, interval, ordinal, binary

keep : True or False

dtype : int, float, str

In [13]:
train_df.columns

Index(['id', 'target', 'ps_ind_01', 'ps_ind_02_cat', 'ps_ind_03',
       'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_ind_06_bin', 'ps_ind_07_bin',
       'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin', 'ps_ind_11_bin',
       'ps_ind_12_bin', 'ps_ind_13_bin', 'ps_ind_14', 'ps_ind_15',
       'ps_ind_16_bin', 'ps_ind_17_bin', 'ps_ind_18_bin', 'ps_reg_01',
       'ps_reg_02', 'ps_reg_03', 'ps_car_01_cat', 'ps_car_02_cat',
       'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat', 'ps_car_06_cat',
       'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat', 'ps_car_10_cat',
       'ps_car_11_cat', 'ps_car_11', 'ps_car_12', 'ps_car_13', 'ps_car_14',
       'ps_car_15', 'ps_calc_01', 'ps_calc_02', 'ps_calc_03', 'ps_calc_04',
       'ps_calc_05', 'ps_calc_06', 'ps_calc_07', 'ps_calc_08', 'ps_calc_09',
       'ps_calc_10', 'ps_calc_11', 'ps_calc_12', 'ps_calc_13', 'ps_calc_14',
       'ps_calc_15_bin', 'ps_calc_16_bin', 'ps_calc_17_bin', 'ps_calc_18_bin',
       'ps_calc_19_bin', 'ps_calc_20_bin'],


In [14]:
# metadata 분류 작업
data = []
for f in train_df.columns :
    # role 정의하기
    if f == "target" :
        role = "target"
    elif f == "id" :
        role = "id"
    else :
        role = "input"
        
    # level 정의하기
    if ("bin" in f) or (f == "target") :
        level = "binary"
    elif ("cat" in f) or f =="id" :
        level = "nominal"
    elif train_df[f].dtype == float :
        level = "interval"
    elif train_df[f].dtype == int :
        level = "ordinal"
        
    # id값만 제외하고 모든 값들에 keep 을 True로 설정
    keep = True
    if f == "id" :
        keep = False
    
    # 데이터타입을 정의
    dtype = train_df[f].dtype
    
    # 위에서 구한 메타데이터를 하나의 딕셔너리로 담는다.
    f_dict = {
        "varname": f,
        "role" : role,
        "level" : level,
        "keep" : keep,
        "dtype" : dtype
    }
    data.append(f_dict)

In [15]:
meta = pd.DataFrame(data, columns=["varname", "role", "level", "keep", "dtype"])
meta.set_index("varname", inplace=True)
meta

role     level   keep    dtype
varname                                         
id                  id   nominal  False    int64
target          target    binary   True    int64
ps_ind_01        input    binary   True    int64
ps_ind_02_cat    input   nominal   True    int64
ps_ind_03        input   nominal   True    int64
ps_ind_04_cat    input   nominal   True    int64
ps_ind_05_cat    input   nominal   True    int64
ps_ind_06_bin    input    binary   True    int64
ps_ind_07_bin    input    binary   True    int64
ps_ind_08_bin    input    binary   True    int64
ps_ind_09_bin    input    binary   True    int64
ps_ind_10_bin    input    binary   True    int64
ps_ind_11_bin    input    binary   True    int64
ps_ind_12_bin    input    binary   True    int64
ps_ind_13_bin    input    binary   True    int64
ps_ind_14        input    binary   True    int64
ps_ind_15        input    binary   True    int64
ps_ind_16_bin    input    binary   True    int64
ps_ind_17_bin    input    binary   True    int64
ps_ind_18_bin    input    binary   True    int64
ps_reg_01        input  interval   True  float64
ps_reg_02        input  interval   True  float64
ps_reg_03        input  interval   True  float64
ps_car_01_cat    input   nominal   True    int64
ps_car_02_cat    input   nominal   True    int64
ps_car_03_cat    input   nominal   True    int64
ps_car_04_cat    input   nominal   True    int64
ps_car_05_cat    input   nominal   True    int64
ps_car_06_cat    input   nominal   True    int64
ps_car_07_cat    input   nominal   True    int64
ps_car_08_cat    input   nominal   True    int64
ps_car_09_cat    input   nominal   True    int64
ps_car_10_cat    input   nominal   True    int64
ps_car_11_cat    input   nominal   True    int64
ps_car_11        input   nominal   True    int64
ps_car_12        input  interval   True  float64
ps_car_13        input  interval   True  float64
ps_car_14        input  interval   True  float64
ps_car_15        input  interval   True  float64
ps_calc_01       input  interval   True  float64
ps_calc_02       input  interval   True  float64
ps_calc_03       input  interval   True  float64
ps_calc_04       input  interval   True    int64
ps_calc_05       input  interval   True    int64
ps_calc_06       input  interval   True    int64
ps_calc_07       input  interval   True    int64
ps_calc_08       input  interval   True    int64
ps_calc_09       input  interval   True    int64
ps_calc_10       input  interval   True    int64
ps_calc_11       input  interval   True    int64
ps_calc_12       input  interval   True    int64
ps_calc_13       input  interval   True    int64
ps_calc_14       input  interval   True    int64
ps_calc_15_bin   input    binary   True    int64
ps_calc_16_bin   input    binary   True    int64
ps_calc_17_bin   input    binary   True    int64
ps_calc_18_bin   input    binary   True    int64
ps_calc_19_bin   input    binary   True    int64
ps_calc_20_bin   input    binary   True    int64

In [16]:
#예시로, 숫자 데이터가 포함되어 있고 dropped 되지 않은 열을 불러온다.

meta[(meta.level == "nominal") & (meta.keep)].index

Index(['ps_ind_02_cat', 'ps_ind_03', 'ps_ind_04_cat', 'ps_ind_05_cat',
       'ps_car_01_cat', 'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_04_cat',
       'ps_car_05_cat', 'ps_car_06_cat', 'ps_car_07_cat', 'ps_car_08_cat',
       'ps_car_09_cat', 'ps_car_10_cat', 'ps_car_11_cat', 'ps_car_11'],
      dtype='object', name='varname')

In [17]:
#아래는 role과 level이 총 몇개가 들어있는지에 대한 데이터프레임

pd.DataFrame({"count" : meta.groupby(["role", "level"])["role"].count()})

count
role   level          
id     nominal       1
input  binary       20
       interval     21
       nominal      16
target binary        1

In [18]:
pd.DataFrame({"count" : meta.groupby(["role", "level"])["role"].count()}).reset_index()

role     level  count
0      id   nominal      1
1   input    binary     20
2   input  interval     21
3   input   nominal     16
4  target    binary      1

# 통계 분석 
보통 통계 분석을 할때는 describe()함수를 사용하지만, mean, std, 등을 카테고리나 id 값에 사용하는 것은 적절하지 않다. 

우리는 그 대신 아까 구한 메타 데이터를 이용한다. 메타 데이터 덕분에 우리는 describe()를 사용하고 싶은 칼럼만 더 쉽게 구할 수 있다.

In [19]:
# 구간 값 변수
v = meta[ (meta.level == "ordinal") & (meta.keep)].index
v
# 구간값이 있는 칼럼의 이름만 구했다.

Index([], dtype='object', name='varname')

- min을 잘 살펴보자.
- min이 -1이 있는 칼럼을 나열한다.

| reg 변수 | car 변수 |calc 변수 |
|--------|--------|--------|
|ps_reg_03가 결측치 포함 | ps_car_12, ps_car_14가 결측치를 포함 | 결측치 없음 |
| 각 변수들의 최소-최대값의 범위는 굉장히 다양하다. 정규화를 사용할 수 있지만, 정규화는 classifier에 달려있음 | 스케일링 사용할것 | 세 calc 최대값이 모두 0.9|
|x  | x| 세 calc 변수들은 비슷한 분포를 가지고 있음|
                                       

=> 전체적으로, 범위 변수들의 범위가 상당히 작음

=> 아마도 로그화나 다른 데이터 변형 작업이 실행되었을 가능성이 있음

# 이진 변수들

In [22]:
v = meta[ (meta.level == "binary") & (meta.keep)].index
train_df[v].describe()

target      ps_ind_01  ps_ind_06_bin  ps_ind_07_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.036448       1.900378       0.393742       0.257033   
std         0.187401       1.983789       0.488579       0.436998   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         0.000000       3.000000       1.000000       1.000000   
max         1.000000       7.000000       1.000000       1.000000   

       ps_ind_08_bin  ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.163921       0.185304       0.000373       0.001692   
std         0.370205       0.388544       0.019309       0.041097   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_ind_12_bin  ps_ind_13_bin      ps_ind_14      ps_ind_15  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.009439       0.000948       0.012451       7.299922   
std         0.096693       0.030768       0.127545       3.546042   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       5.000000   
50%         0.000000       0.000000       0.000000       7.000000   
75%         0.000000       0.000000       0.000000      10.000000   
max         1.000000       1.000000       4.000000      13.000000   

       ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_calc_15_bin  \
count  595212.000000  595212.000000  595212.000000   595212.000000   
mean        0.660823       0.121081       0.153446        0.122427   
std         0.473430       0.326222       0.360417        0.327779   
min         0.000000       0.000000       0.000000        0.000000   
25%         0.000000       0.000000       0.000000        0.000000   
50%         1.000000       0.000000       0.000000        0.000000   
75%         1.000000       0.000000       0.000000        0.000000   
max         1.000000       1.000000       1.000000        1.000000   

       ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
count   595212.000000   595212.000000   595212.000000   595212.000000   
mean         0.627840        0.554182        0.287182        0.349024   
std          0.483381        0.497056        0.452447        0.476662   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          1.000000        1.000000        0.000000        0.000000   
75%          1.000000        1.000000        1.000000        1.000000   
max          1.000000        1.000000        1.000000        1.000000   

       ps_calc_20_bin  
count   595212.000000  
mean         0.153318  
std          0.360295  
min          0.000000  
25%          0.000000  
50%          0.000000  
75%          0.000000  
max          1.000000

이진변수이므로 모두 0,1만 들어있다.

- target이 포함되어있는 이진 변수들의 평균은 3.645%로, 매우 불균형적으로 구성되어 있다.
- 낮은 평균을 고려해보면 대부분의 경우 대부분의 타겟값들은 0이라고 할 수 있다.

불균형 클래스 조정하기 
위에서 설명했듯, 타겟 칼럼이 1인 비율은 0인 비율에 비해 훨씬 낮은 것을 알 수 있다. 이 결론을 바탕으로 모델을 만들때 꽤 정확하게 예측하는 모델을 만들 수는 있지만, 다른 추가된 값을 실질적으로 포함할 수 있다. 이것을 바탕으로, 2개의 방법을 고려해볼수 있겠다.

- target이 1인 항목을 oversampling 하는것.
- target이 0인 항목을 undersampling 하는것.

현재 트레이닝 모델에 있는 데이터 셋이 많으므로, target이 0인 항목을 undersampling을 하는것으로 결정한다.

In [23]:
desired_apriori = 0.10

# 타겟의 인덱스 구하기
idx_0 = train_df.query("target == 0").index
idx_1 = train_df.query("target == 1").index

idx_0

Int64Index([     0,      1,      2,      3,      4,      5,      6,      7,
                 8,     10,
            ...
            595202, 595203, 595204, 595205, 595206, 595207, 595208, 595209,
            595210, 595211],
           dtype='int64', length=573518)

In [24]:
idx_1

Int64Index([     9,     19,     28,     39,     41,     65,     87,    108,
               191,    224,
            ...
            594968, 594975, 594980, 594982, 594985, 595006, 595046, 595099,
            595113, 595158],
           dtype='int64', length=21694)

target이 0인 행이 573518개인 반면, target이 1인 행은 21694개

In [25]:
# 각 행의 갯수 저장
nb_0 = len(idx_0)
nb_1 = len(idx_1)

In [26]:
# undersampling을 할 비율
undersample_rate = ((1-desired_apriori) *nb_1) / (nb_0 * desired_apriori)
undersampled_nb_0 = int(undersample_rate * nb_0)
print(f"언더샘플을 할 비율 : {undersample_rate}")
print(f"언더샘플 후에 target 0의 개수 : {undersampled_nb_0}")

언더샘플을 할 비율 : 0.34043569687437886
언더샘플 후에 target 0의 개수 : 195246


In [27]:
# 목표 priori에 도달하기 위해 언더샘플된 랜덤으로 target=0 행의 인덱스의 순서를 섞는다.
undersampled_idx = shuffle(idx_0, random_state=37, n_samples=undersampled_nb_0)

# 두 인덱스들을 합쳐 하나의 리스트로 저장
idx_list = list(undersampled_idx) + list(idx_1) 

# undersample된 데이터프레임을 새롭게 저장한다.
train_df = train_df.loc[idx_list].reset_index(drop=True)
train_df

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0        353762       0          1              1          1              1   
1        845629       0          5              1          7              1   
2       1449185       0          4              1          3              1   
3        834066       0          6              1          6              0   
4        436189       0          5              1          5              0   
...         ...     ...        ...            ...        ...            ...   
216935  1487473       1          5              1          7              0   
216936  1487566       1          1              1          5              0   
216937  1487716       1          0              1          6              1   
216938  1487748       1          0              2          2              0   
216939  1487866       1          1              2          1              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
0                   0              1              0              0   
1                   1              1              0              0   
2                   0              1              0              0   
3                   0              1              0              0   
4                   0              0              0              1   
...               ...            ...            ...            ...   
216935              3              0              0              1   
216936              0              0              0              1   
216937              0              0              1              0   
216938              0              1              0              0   
216939              0              0              0              1   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
0                   0              0              0              0   
1                   0              0              0              0   
2                   0              0              0              0   
3                   0              0              0              0   
4                   0              0              0              0   
...               ...            ...            ...            ...   
216935              0              0              0              0   
216936              0              0              0              0   
216937              0              0              0              0   
216938              0              0              0              0   
216939              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
0                   0          0         11              0              1   
1                   0          0          5              0              0   
2                   0          0          4              1              0   
3                   0          0          8              1              0   
4                   0          0         10              1              0   
...               ...        ...        ...            ...            ...   
216935              0          0          8              0              1   
216936              0          0          4              0              0   
216937              0          0          9              0              1   
216938              0          0          2              0              0   
216939              0          0         13              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
0                   0        0.6        0.6   0.838153             11   
1                   1        0.9        0.6   0.728440              7   
2                   0        0.9        0.6   0.869267             11   
3                   0        0.6        1.5   1.705872              6   
4                   0        0.8        0.8   1.086566              5   
...               .

In [28]:
train_df["target"].value_counts()

0    195246
1     21694
Name: target, dtype: int64